<launch>  
  <arg name="localization"      default="false"/>
  <arg name="database_path"     default="rtabmap.db"/>
  <arg     if="$(arg localization)" name="args"  default=""/>
  <arg unless="$(arg localization)" name="args"  default="--delete_db_on_start"/>
  <arg name="preview" default="false" />

  <!-- <node name="kinect1_broadcaster" pkg="tf" type="static_transform_publisher" args="0.15 0 0.3 0 0 0 1 base_link camera_link 100" /> -->
  
  <node name="pointcloud_to_laserscan_node" pkg="pointcloud_to_laserscan" type="pointcloud_to_laserscan_node" >
    <param name="target_frame" value="camera_link" />
    <param name="min_height" value="-0.0"/> 
    <param name="angle_min" value="-pi/2"/>
    <param name="angle_max" value="pi/2"/>  
    <remap from="cloud_in" to="/camera/depth_registered/points" />
    <remap from="scan" to="scan" />
  </node>

  <include file="$(find openni_launch)/launch/openni.launch" unless="$(arg preview)">
    <arg name="depth_registration" value="true"/>
  </include>

  <group ns="rtabmap">

    <node pkg="nodelet" type="nodelet" name="rgbd_sync" args="standalone rtabmap_ros/rgbd_sync" output="screen">
      <remap from="rgb/image"       to="/camera/rgb/image_rect_color"/>
      <remap from="depth/image"     to="/camera/depth_registered/image_raw"/>
      <remap from="rgb/camera_info" to="/camera/rgb/camera_info"/>
      <remap from="rgbd_image"      to="rgbd_image"/> <!-- output -->
      
      <!-- Should be true for not synchronized camera topics 
           (e.g., false for kinectv2, zed, realsense, true for xtion, kinect360)-->
      <param name="approx_sync"       value="true"/> 
    </node>

    <node name="rtabmap" pkg="rtabmap_ros" type="rtabmap" output="screen" args="$(arg args)">
      <param name="frame_id" type="string" value="base_link"/>
      <param name="database_path"       type="string" value="$(arg database_path)"/>
      <param name="subscribe_depth" type="bool" value="false"/>
      <param name="subscribe_rgbd" type="bool" value="true"/>

      <remap from="odom" to="/odom"/>

      <remap from="rgbd_image" to="rgbd_image"/>
      <remap from="grid_map" to="/map" />
      <param name="queue_size" type="int" value="10"/>

      <!-- RTAB-Map's parameters -->
      <param name="RGBD/AngularUpdate"        type="string" value="0.01"/>
      <param name="RGBD/LinearUpdate"         type="string" value="0.01"/>
      <param name="RGBD/OptimizeFromGraphEnd" type="string" value="false"/>

      <!-- localization mode -->
      <param     if="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="false"/>
      <param unless="$(arg localization)" name="Mem/IncrementalMemory" type="string" value="true"/>
      <param name="Mem/InitWMWithAllNodes" type="string" value="$(arg localization)"/> 
    </node>
  </group>
</launch>